File: ts/src/pro/test/Exchange/test.watchOrderBook.ts

import testOrderBook from '../../../test/Exchange/base/test.orderBook.js';
import testSharedMethods from '../../../test/Exchange/base/test.sharedMethods.js';
import { InvalidNonce } from '../../../base/errors.js';
import { Exchange } from '../../../../ccxt.js';
import type { OrderBook } from '../../../base/types.js';

async function testWatchOrderBook (exchange: Exchange, skippedProperties: object, symbol: string) {
    const method = 'watchOrderBook';
    // `watchOrderBook` only resolves when the exchange pushes an update, and a
    // pending subscription can not be cancelled from here, so every extra
    // iteration risks blocking until the test-runner kills the whole exchange.
    // a validated book is already a pass, so keep sampling only while updates
    // keep arriving quickly and stop once the book goes quiet.
    const maxIdleTime = 5000;
    let now = exchange.milliseconds ();
    const ends = now + 15000;
    let idle = false;
    while ((now < ends) && !idle) {
        let response: OrderBook | undefined = undefined;
        let success = true;
        const startTime = exchange.milliseconds ();
        try {
            response = await exchange.watchOrderBook (symbol);
        } catch (e) {
            if (!testSharedMethods.isTemporaryFailure (e) && !(e instanceof InvalidNonce)) {
                throw e;
            }
            success = false;
        }
        // refresh the deadline on every path, otherwise a stream of temporary
        // failures would loop forever
        now = exchange.milliseconds ();
        if ((success === true) && (response !== undefined)) {
            testOrderBook (exchange, skippedProperties, method, response, symbol);
            const elapsed = now - startTime;
            if (elapsed > maxIdleTime) {
                // this market updates slower than the remaining test window, so
                // awaiting another delta would only end in a harness timeout
                idle = true;
            }
        }
    }
    return true;
}

export default testWatchOrderBook;

File: julia/Ccxt/test/pro/exchange/test.watchOrderBook.jl
using Test
using Ccxt
function testWatchOrderBook(exchange, skippedProperties, symbol)

    method = "watchOrderBook";
    now = milliseconds(exchange);
    ends = now + 15000;
    while functions.ccxtruthy(functions.ccxt_lt(now, ends))
        response = nothing;
        success = true;
        try
            response = Base.fetch(watchOrderBook(exchange, symbol));
        catch e
            if functions.ccxtruthy(!functions.ccxtruthy(isTemporaryFailure(testSharedMethods, e)))
                throw(e);
            end
            now = milliseconds(exchange);
            success = false;

        end
        if functions.ccxtruthy(@functions.ccxt_and((success), (response != nothing)))
            @test functions.ccxtruthy(isDictionary(exchange, response))
            now = milliseconds(exchange);
            testOrderBook(exchange, skippedProperties, method, response, symbol);
        end
    end
    return true
end
